cmd/ddev/cmd/hostname.go

```diff
@@ -5,58 +5,235 @@ import (
+to allow ddev to modify your hosts file.`,
 	Run: func(cmd *cobra.Command, args []string) {
-		if len(args) != 2 {
-			output.UserOut.Fatal("Invalid arguments supplied. Please use 'ddev hostname [hostname] [ip]'")
-		}
-
-		rawResult := make(map[string]interface{})
-
-		hostname, ip := args[0], args[1]
 		hosts, err := goodhosts.NewHosts()
 		if err != nil {
+			rawResult := make(map[string]interface{})
+			detail := fmt.Sprintf("Could not open hosts file for reading: %v", err)
 			rawResult["error"] = "READERROR"
-			rawResult["full_error"] = fmt.Sprintf("%v", err)
-			output.UserOut.WithField("raw", rawResult).Fatal(fmt.Sprintf("could not open hosts file for read: %v", err))
+			rawResult["full_error"] = detail
+			output.UserOut.WithField("raw", rawResult).Fatal(detail)
+
+			return
+		}
+
+		// Attempt to write the hosts file first to catch any permissions issues early
+		if err := hosts.Flush(); err != nil {
+			rawResult := make(map[string]interface{})
+			detail := fmt.Sprintf("Could not write hosts file: %v", err)
+			rawResult["error"] = "WRITEERROR"
+			rawResult["full_error"] = detail
+			output.UserOut.WithField("raw", rawResult).Fatal(detail)
+
 			return
 		}
-		if hosts.Has(ip, hostname) {
-			if output.JSONOutput {
-				rawResult["error"] = "SUCCESS"
-				rawResult["detail"] = "hostname already exists in hosts file"
-				output.UserOut.WithField("raw", rawResult).Info("")
+
+		// If requested, remove all inactive host names and exit
+		if removeInactive {
+			if len(args) > 0 {
+				output.UserOut.Fatal("Invalid arguments supplied. 'ddev hostname --remove-all' accepts no arguments.")
 			}
+
+			removeInactiveHostnames(hosts)
+
 			return
 		}
 
-		err = hosts.Add(ip, hostname)
-		if err != nil {
-			rawResult["error"] = "ADDERROR"
-			rawResult["full_error"] = fmt.Sprintf("%v", err)
-			output.UserOut.WithField("raw", rawResult).Fatal(fmt.Sprintf("could not add hostname %s at %s: %v", hostname, ip, err))
+		// If operating on one host name, two arguments are required
+		if len(args) != 2 {
+			output.UserOut.Fatal("Invalid arguments supplied. Please use 'ddev hostname [hostname] [ip]'")
 		}
 
-		if err := hosts.Flush(); err != nil {
-			rawResult["error"] = "WRITEERROR"
-			rawResult["full_error"] = fmt.Sprintf("%v", err)
-			output.UserOut.WithField("raw", rawResult).Fatal(fmt.Sprintf("Could not write hosts file: %v", err))
-		} else {
-			rawResult["error"] = "SUCCESS"
-			rawResult["detail"] = "hostname added to hosts file"
-			output.UserOut.WithField("raw", rawResult).Info("")
+		hostname, ip := args[0], args[1]
+
+		// If requested, remove the provided host name and exit
+		if removeHostName {
+			removeHostname(hosts, ip, hostname)
+
+			return
 		}
+
+		// By default, add a host name
+		addHostname(hosts, ip, hostname)
 	},
 }
 
+// addHostname encapsulates the logic of adding a hostname to the system's hosts file.
+func addHostname(hosts goodhosts.Hosts, ip, hostname string) {
```

Comment:
Seems like a relic of long ago that all this has remained in cmd instead of getting refactored into pkg. OK with me, and this is a very, very isolated command.